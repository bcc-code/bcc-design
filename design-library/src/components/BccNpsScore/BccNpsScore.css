@layer components {
  .bcc-nps-score {
    @apply min-w-[280px] p-4 text-sm;
    @apply flex flex-col gap-y-2;
  }

  .bcc-nps-score--heading {
    @apply flex items-center justify-between;
  }

  .bcc-nps-score--label {
    @apply flex-1 truncate;
  }

  .bcc-nps-score--bar {
    @apply flex justify-evenly rounded-xl text-white;
    @apply bg-gradient-to-r from-red-700 via-yellow-400 via-75% to-green-500;
  }
  .bcc-nps-score--bar:not(.disabled) {
    @apply pointer-events-none cursor-pointer;
  }

  .bcc-nps-score--bar.reverse {
    @apply bg-gradient-to-l;
  }

  .bcc-nps-score--container {
    @apply flex flex-1;
  }

  .bcc-nps-score--number {
    @apply flex h-8 flex-1 items-center justify-center;
    @apply pointer-events-auto font-bold transition-all duration-200 hover:bg-transparent;
  }

  .bcc-nps-score--number.active {
    @apply pointer-events-none;
  }

  .bcc-nps-score--number.inactive {
    @apply bg-neutral-200 text-black;
  }

  .bcc-nps-score--container:first-child .bcc-nps-score--number {
    @apply rounded-l-xl;
  }
  .bcc-nps-score--container:last-child .bcc-nps-score--number {
    @apply rounded-r-xl;
  }

  .bcc-nps-score--container .separator {
    @apply h-full w-0.5 bg-white;
  }
}
