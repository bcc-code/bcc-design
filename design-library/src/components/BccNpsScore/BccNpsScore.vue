<script setup lang="ts">
import { computed } from "vue";

type Props = {
  reverse?: boolean;
  leftLabel?: string;
  rightLabel?: string;
  labelPosition?: "top" | "bottom";
  min?: number;
  max?: number;
  disabled?: boolean;
};

const modelValue = defineModel<number | null>({
  default: null,
});

const props = withDefaults(defineProps<Props>(), {
  reverse: false,
  disabled: false,
  leftLabel: "Not likely",
  rightLabel: "Very likely",
  labelPosition: "top",
  min: 0,
  max: 10,
});

let anim: NodeJS.Timeout | null = null;
function selected(i: number) {
  if (props.disabled) return;

  if (modelValue.value === null) {
    modelValue.value = 0;
  }
  anim && clearInterval(anim);
  anim = setInterval(() => {
    if (modelValue.value === i) {
      return anim && clearInterval(anim);
    }
    if (modelValue.value! < i) {
      ++modelValue.value!;
    } else {
      --modelValue.value!;
    }
  }, 20);
}

const range = computed(() => {
  let result = [];
  for (let i = props.min; i <= props.max; i++) {
    result.push(i);
  }
  return result;
});
</script>

<template>
  <div class="bcc-nps-score">
    <div
      class="bcc-nps-score--heading"
      :class="labelPosition === 'top' ? 'order-1' : labelPosition === 'bottom' ? 'order-last' : ''"
    >
      <p class="bcc-nps-score--label">{{ leftLabel }}</p>
      <p class="bcc-nps-score--label text-right">{{ rightLabel }}</p>
    </div>
    <div class="bcc-nps-score--bar order-2" :class="{ disabled, reverse }">
      <button
        v-for="num in range"
        :key="num"
        @click="selected(num)"
        :disabled="disabled"
        class="bcc-nps-score--number"
        :class="{
          inactive: modelValue === null || isNaN(modelValue) || modelValue < num,
          active: modelValue === num,
        }"
      >
        {{ num }}
      </button>
    </div>
  </div>
</template>
