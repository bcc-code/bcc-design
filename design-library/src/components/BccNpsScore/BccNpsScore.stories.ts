import { ref } from "vue";
import BccNpsScore from "./BccNpsScore.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

export default {
  title: "Feedback/BccNpsScore",
  component: BccNpsScore,
  argTypes: {
    reverse: {
      description: "Should the bar be reversed?",
    },
    labelPosition: {
      description: "Where to place the labels",
      options: ["top", "bottom"],
      control: { type: "radio" },
    },
  },
} as Meta<typeof BccNpsScore>;

const Template: StoryFn<typeof BccNpsScore> = (args) => ({
  components: { BccNpsScore },
  setup() {
    const value = ref(0);
    return { args, value };
  },
  template: `
    <BccNpsScore v-bind="args" v-model="value" style="width: 320px" />
  `,
});

export const Example = Template.bind({});
Example.args = {};

const TemplateForLongRange: StoryFn<typeof BccNpsScore> = (args) => ({
  components: { BccNpsScore },
  setup() {
    const value = ref(0);
    return { args, value };
  },
  template: `
    <BccNpsScore v-bind="args" v-model="value" style="width: 992px" />
  `,
});

export const ExampleForLongRang = TemplateForLongRange.bind({});
ExampleForLongRang.args = {
  max: 30,
  labelPosition: "bottom",
};

const TemplateForInitialValue: StoryFn<typeof BccNpsScore> = (args) => ({
  components: { BccNpsScore },
  setup() {
    const value = ref(5);
    return { args, value };
  },
  template: `
    <BccNpsScore v-bind="args" v-model="value" style="width: 320px" />
  `,
});

export const ExampleAsDisabled = TemplateForInitialValue.bind({});
ExampleAsDisabled.args = {
  max: 10,
  disabled: true,
};
