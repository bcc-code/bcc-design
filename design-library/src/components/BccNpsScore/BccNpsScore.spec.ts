import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import BccNpsScore from "./BccNpsScore.vue";

describe("BccNpsScore", () => {
  it("renders a nps score", () => {
    expect(BccNpsScore).toBeTruthy();

    const wrapper = mount(BccNpsScore, {
      props: {},
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
