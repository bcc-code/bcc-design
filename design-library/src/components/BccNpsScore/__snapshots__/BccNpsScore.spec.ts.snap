// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccNpsScore > renders a nps score 1`] = `
"<div class=\\"bcc-nps-score\\">
  <div class=\\"bcc-nps-score--heading order-1\\">
    <p class=\\"bcc-nps-score--label\\">Not likely</p>
    <p class=\\"bcc-nps-score--label text-right\\">Very likely</p>
  </div>
  <div class=\\"bcc-nps-score--bar order-2\\"><button class=\\"bcc-nps-score--number inactive\\">0</button><button class=\\"bcc-nps-score--number inactive\\">1</button><button class=\\"bcc-nps-score--number inactive\\">2</button><button class=\\"bcc-nps-score--number inactive\\">3</button><button class=\\"bcc-nps-score--number inactive\\">4</button><button class=\\"bcc-nps-score--number inactive\\">5</button><button class=\\"bcc-nps-score--number inactive\\">6</button><button class=\\"bcc-nps-score--number inactive\\">7</button><button class=\\"bcc-nps-score--number inactive\\">8</button><button class=\\"bcc-nps-score--number inactive\\">9</button><button class=\\"bcc-nps-score--number inactive\\">10</button></div>
</div>"
`;
